torch/csrc/autograd/record_function.cpp

```diff
@@ -18,85 +18,147 @@ float sample_zero_one() {
+    TORCH_CHECK(!cb_list.empty(), "Empty callbacks stack");
+    cb_list.pop_back();
+    if (!is_thread_local) {
+      ++global_callbacks_version_;
+    }
+  }
+
+  inline bool hasGlobalCallbacks() const {
     return !callbacks_.empty();
   }
 
+  inline bool hasThreadLocalCallbacks() const {
+    return !tls_callbacks_.empty();
+  }
+
   inline bool needsInputs() const {
     return has_callbacks_with_inputs_;
   }
 
   void runStartCallbacks(RecordFunction& rf) {
-    rf._setCallbacksVersion(callbacks_version_);
-    rf._activeCallbacks().clear();
-    for (size_t cb_idx = 0; cb_idx < callbacks_.size(); ++cb_idx) {
-      if (shouldRunCallback(cb_idx, rf.scope())) {
-        try {
-          bool cb_ret = callbacks_[cb_idx].start_cb_(rf);
-          rf._activeCallbacks().push_back(cb_ret);
-        } catch (const std::exception &e) {
-          LOG(WARNING) << "Exception in RecordFunction start observer: "
-                       << e.what();
-          rf._activeCallbacks().push_back(false);
-        } catch (...) {
-          LOG(WARNING) << "Exception in RecordFunction start observer: unknown";
-          rf._activeCallbacks().push_back(false);
-        }
+    rf._setGlobalCallbacksVersion(global_callbacks_version_);
+    rf._activeGlobalCallbacks().clear();
+    for (auto& cb : callbacks_) {
+      if (shouldRunCallback(cb, rf.scope())) {
+        auto ret = tryRunCallback(cb.start(), rf);
+        rf._activeGlobalCallbacks().push_back(ret);
       } else {
-        rf._activeCallbacks().push_back(false);
+        rf._activeGlobalCallbacks().push_back(false);
+      }
+    }
+    for (auto& cb : tls_callbacks_) {
+      if (cb.checkScope(rf.scope())) {
+        tryRunCallback(cb.start(), rf);
       }
     }
   }
 
   void runEndCallbacks(RecordFunction& rf) {
-    if (rf._callbacksVersion() == callbacks_version_) {
-      for (size_t cb_idx = 0; cb_idx < rf._activeCallbacks().size(); ++cb_idx) {
-        if (!rf._activeCallbacks()[cb_idx]) {
+    if (rf._globalCallbacksVersion() == global_callbacks_version_) {
+      TORCH_CHECK(rf._activeGlobalCallbacks().size() == callbacks_.size());
+      for (size_t cb_idx = 0; cb_idx < rf._activeGlobalCallbacks().size(); ++cb_idx) {
+        if (!rf._activeGlobalCallbacks()[cb_idx]) {
           continue;
         }
-        try {
-          callbacks_[cb_idx].end_cb_(rf);
-        } catch (const std::exception &e) {
-          LOG(WARNING) << "Exception in RecordFunction end observer: "
-                       << e.what();
-        } catch (...) {
-          LOG(WARNING) << "Exception in RecordFunction end observer: unknown";
-        }
+        tryRunCallback(callbacks_[cb_idx].end(), rf);
       }
     } else {
-      LOG(WARNING) << "Callbacks changed while running a record function, "
-                   << "you might be partially overlapping a record function "
-                   << "with a profiling scope";
+      VLOG(1) << "Global callbacks changed while running a record function, "
```

Comment:
I think this is important enough to stay as the more verbose `LOG`, what do you think? This warning helped me debug/discover a few race conditions with the profiler + RPC. Also, seeing this warning means that the `RecordFunction` never had it's end method called, so it won't show up correctly in the profiler output. 

Additionally, should we add `rf.name()` in this message to help with the debug?

Reply:
sure, will upd